apiserver/remotefirewaller/ingressaddresswatcher.go

```diff
@@ -156,36 +198,47 @@ func (w *IngressAddressWatcher) unitAddress(unit Unit) (string, bool, error) {
 	return addr.Value, true, nil
 }
 
-func (w *IngressAddressWatcher) mergeAddresses(addresses set.Strings, c params.RelationUnitsChange) (set.Strings, error) {
-	result := set.NewStrings(addresses.Values()...)
+func (w *IngressAddressWatcher) relationUnitsChanged(c params.RelationUnitsChange) (bool, error) {
+	changed := false
 	for name := range c.Changed {
 
 		u, err := w.backend.Unit(name)
 		if errors.IsNotFound(err) {
 			continue
 		}
 		if err != nil {
-			return result, err
+			return false, err
+		}
+
+		if err := w.trackUnit(u); err != nil {
+			return false, errors.Trace(err)
 		}
 
-		// TODO - start watcher to pick up machine address changes
 		// We need to know whether to look at the public or cloud local address.
 		// For now, we'll use the public address and later if needed use a watcher
 		// parameter to look at the cloud local address.
 		addr, ok, err := w.unitAddress(u)
 		if err != nil {
-			return result, err
+			return false, err
 		}
 		if !ok {
 			continue
 		}
-		result.Add(addr)
-		w.known[name] = addr
+		if w.known[name] != addr {
+			w.known[name] = addr
+			changed = true
+		}
 	}
 	for _, name := range c.Departed {
+		if err := w.untrackUnit(name); err != nil {
+			return false, errors.Trace(err)
+		}
 		// If the unit is departing and we have seen its address,
 		// remove the address.
-		address := w.known[name]
+		address, ok := w.known[name]
+		if !ok {
```

Comment:
I'm using `ok` to prevent reporting we've changed the map when the value wasn't in there to begin with - it's really there to stop the loop further down from happening.